### PR TITLE
UCP/LISTENER: ucp_listener_create

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
## What
update doc and fix param check

## Why ?
fixes https://github.com/openucx/ucx/issues/6257, see conversation https://github.com/openucx/ucx/issues/6257#issuecomment-771981284

### CODE REVIEW

File: src/ucp/core/ucp_listener.c [hunk 1]
```diff
@@ -230,10 +232,9 @@ ucs_status_t ucp_listener_create(ucp_worker_h worker,
 
     UCP_CHECK_PARAM_NON_NULL(params->sockaddr.addr, status, return status);
 
-    if (ucs_test_all_flags(params->field_mask,
-                           UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER |
-                           UCP_LISTENER_PARAM_FIELD_CONN_HANDLER)) {
-        ucs_error("only one accept handler should be provided");
+    if (ucs_test_all_flags(params->field_mask, handlers_mask) ||
+        (!(params->field_mask & handlers_mask))) {
```
Comment (dmitrygx) [inline src/ucp/core/ucp_listener.c:236]: excessive `()`

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 2]
```diff
@@ -758,6 +762,52 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle) {
...
+    ucp_listener_h listener;
+    ucs_status_t status;
+
+    params.field_mask = 0;
+    /* address and accept handlers are not specified */
+    status = ucp_listener_create(receiver().worker(), &params, &listener);
+    ASSERT_EQ(UCS_ERR_INVALID_PARAM, status);
+
+    /* add listen addr */
+    params.field_mask |= UCP_LISTENER_PARAM_FIELD_SOCK_ADDR;
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_sockaddr.cc:779]: not aligned by `=`

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 3]
```diff
@@ -758,6 +762,52 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle) {
     EXPECT_EQ(1u, sender().get_err_num());
 }
 
+UCS_TEST_SKIP_COND_P(test_ucp_sockaddr, listener_invalid_params,
+                     nonparameterized_test())
+{
+    scoped_log_handler wrap_err(wrap_errors_logger);
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_sockaddr.cc:768]: add wrapping errors only where we expect `ucs_error`?
e.g.
```
{
    scoped_log_handler wrap_err(wrap_errors_logger);
    /* define handlers to NULL */
    params.field_mask      |= UCP_LISTENER_PARAM_FIELD_CONN_HANDLER;
    params.conn_handler.cb  = NULL;
    params.conn_handler.arg = NULL;

    status = ucp_listener_create(receiver().worker(), &params, &listener);
    ASSERT_EQ(UCS_ERR_INVALID_PARAM, status);
}
...
{
    
}
...
```

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 4]
```diff
@@ -758,6 +762,52 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle) {
...
+    /* add listen addr */
+    params.field_mask |= UCP_LISTENER_PARAM_FIELD_SOCK_ADDR;
+    params.sockaddr.addr    = m_test_addr.get_sock_addr_ptr();
+    params.sockaddr.addrlen = m_test_addr.get_addr_size();
+
+    /* accept handlers aren't set */
+    status = ucp_listener_create(receiver().worker(), &params, &listener);
+    ASSERT_EQ(UCS_ERR_INVALID_PARAM, status);
+
+    /* define handler to NULL */
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_sockaddr.cc:787]: `handlers`

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 5]
```diff
@@ -758,6 +762,52 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle) {
...
+    params.field_mask      |= UCP_LISTENER_PARAM_FIELD_CONN_HANDLER;
+    params.conn_handler.cb  = NULL;
+    params.conn_handler.arg = NULL;
+
+    status = ucp_listener_create(receiver().worker(), &params, &listener);
+    ASSERT_EQ(UCS_ERR_INVALID_PARAM, status);
+
+    /* define 2nd handler */
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_sockaddr.cc:795]: `/* define accept handler */`?

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 6]
```diff
@@ -758,6 +762,52 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle) {
...
+    params.field_mask        |= UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER;
+    params.accept_handler.cb  = NULL;
+    params.accept_handler.arg = NULL;
+
+    status = ucp_listener_create(receiver().worker(), &params, &listener);
+    ASSERT_EQ(UCS_ERR_INVALID_PARAM, status);
+
+    /* reset 2d handler and define 1st, now it's OK */
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_sockaddr.cc:803]: `/* reset accept handler and define connection handler, now it's OK */`?

**Review by dmitrygx — CHANGES_REQUESTED** [review]

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 4 above)
Comment (evgeny-leksikov) [reply test/gtest/ucp/test_ucp_sockaddr.cc:787]: there is 1 handler and its arg

**Comment by yosefe** [comment]
bot:pipe:retest

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 7]
```diff
@@ -758,6 +769,50 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle) {
...
+    ASSERT_EQ(UCS_ERR_INVALID_PARAM, status);
+
+    /* add listen address, use ANY addr to avoid BUSY error in the end */
+    m_test_addr.reset_to_any();
+    params.field_mask      |= UCP_LISTENER_PARAM_FIELD_SOCK_ADDR;
+    params.sockaddr.addr    = m_test_addr.get_sock_addr_ptr();
+    params.sockaddr.addrlen = m_test_addr.get_addr_size();
+    /* accept handlers aren't set */
+    status = create_listener_wrap_err(params, listener);
+    ASSERT_EQ(UCS_ERR_INVALID_PARAM, status);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc:791]: EXPECT_EQ (in other places as well)

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 8]
```diff
@@ -758,6 +769,50 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle) {
...
+    params.conn_handler.arg = NULL;
+    status = create_listener_wrap_err(params, listener);
+    ASSERT_EQ(UCS_ERR_INVALID_PARAM, status);
+
+    /* define accept handler to NULL */
+    params.field_mask        |= UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER;
+    params.accept_handler.cb  = NULL;
+    params.accept_handler.arg = NULL;
+    status = create_listener_wrap_err(params, listener);
+    ASSERT_EQ(UCS_ERR_INVALID_PARAM, status);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc:805]: need to re-set field mask for every test.
and also test with non-NULL cb/arg pointers - to test that combination of 2 handlers, or no handlers at all, causes INVALID_PARAM. and not because the cb is NULL

**Review by yosefe — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@tonycurtis can you pls review API doc change?

**Review by tonycurtis — APPROVED** [review]

**Comment by evgeny-leksikov** [comment]
https://github.com/openucx/ucx/issues/4616
bot:pipe:retest

File: src/ucs/sys/sock.c [hunk 9]
```diff
@@ -391,6 +391,13 @@ ucs_status_t ucs_socket_server_init(const struct sockaddr *saddr, socklen_t sock
         if (status != UCS_OK) {
             goto err_close_socket;
         }
+#ifdef SO_REUSEPORT
```
Comment (yosefe) [inline src/ucs/sys/sock.c:394]: IMO better check it in configure (theoretically, it could be enum/constant and not a macro)
and use `#ifdef HAVE_DECL_SO_REUSEPORT`

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 10]
```diff
@@ -209,7 +209,16 @@ class test_ucp_sockaddr : public ucp_test {
                                           ucp_listener_h &listener)
     {
         scoped_log_handler wrap_err(wrap_errors_logger);
-        return ucp_listener_create(receiver().worker(), &params, &listener);
+
+        /* FIXME: in spite of using SO_REUSEPORT tcp CM can return addr in use error
+         *        for specific port acquired by rdmacm CM */
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc:214]: does it mean tcp can't use same port as rdmacm?
or it means that it happens after rdma_cm already destroyed its id?
given this, do we still need to set SO_REUSEPORT?

Comment (evgeny-leksikov) [reply test/gtest/ucp/test_ucp_sockaddr.cc:214]: TCP can share the port with RDMACM, it means that some port can be free in RDMACM but busy in TCP port space. When RDMACM aquires some port from ANY we shoud start TCP on the same port but it can be busy, for example, by other app

File: src/ucp/core/ucp_listener.c [hunk 11]
```diff
@@ -161,56 +175,68 @@ ucp_listen(ucp_listener_h listener, const ucp_listener_params_t *params)
                                        UCS_SOCKADDR_STRING_LEN),
                       ucs_status_string(status));
 
-            if (status == UCS_ERR_BUSY) {
-                goto err_destroy_listeners;
+            if (status != UCS_ERR_BUSY) {
```
Comment (yosefe) [inline src/ucp/core/ucp_listener.c:178]: maybe flatten the code:
if (status == UCS_OK) {
 ucs_debug(..)
} else if (status == BUSY && use_any_port) {
  /* retry another port */
} else {
  ucs_error
}

File: src/ucp/core/ucp_listener.c [hunk 12]
```diff
@@ -161,56 +175,68 @@ ucp_listen(ucp_listener_h listener, const ucp_listener_params_t *params)
...
+                /* retry another port */
+                status = ucs_sockaddr_set_port(addr, 0);
+                if (status != UCS_OK) {
+                    goto err_free_listeners;
+                }
+
+                ucp_listener_reset_uct_listeners(listener);
+                /* TODO: to reduce probability of "any port busy" need to create
+                 *       TCP listener first */
+                again = 1;
```
Comment (yosefe) [inline src/ucp/core/ucp_listener.c:190]: maybe better:
while (cmpt_index < num_cmpts)
{
  if (ok) { cmpt_index++}
  else if (busy && use_any_port) {
    reset
    cmpt_index = 0;
}

File: src/uct/ib/rdmacm/rdmacm_listener.c [hunk 13]
```diff
@@ -65,7 +65,8 @@ UCS_CLASS_INIT_FUNC(uct_rdmacm_listener_t, uct_cm_h cm,
 
     if (rdma_bind_addr(self->id, (struct sockaddr*)saddr)) {
         status = ((errno == EADDRINUSE) || (errno == EADDRNOTAVAIL)) ?
-                 UCS_ERR_BUSY : UCS_ERR_IO_ERROR;
+                 UCS_ERR_BUSY :
+                 ((errno == ENODEV) ? UCS_ERR_NO_DEVICE : UCS_ERR_IO_ERROR);
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_listener.c:69]: 1. need switch/case here?
2. why need to silence ENODEV?

File: src/ucp/core/ucp_listener.c [hunk 14]
```diff
@@ -181,43 +206,18 @@ ucp_listen(ucp_listener_h listener, const ucp_listener_params_t *params)
...
-            goto err_free_listeners;
-        }
-
-        if (port != uct_listen_port) {
-            ucs_assert(port == 0);
-            status = ucs_sockaddr_set_port(addr, uct_listen_port);
-            if (status != UCS_OK) {
+            log_level = ((status == UCS_ERR_BUSY) ||
+                         (status == UCS_ERR_NO_DEVICE)) ? UCS_LOG_LEVEL_DIAG :
+                        UCS_LOG_LEVEL_ERROR;
```
Comment (yosefe) [inline src/ucp/core/ucp_listener.c:212]: 👍

File: src/uct/ib/rdmacm/rdmacm_listener.c (hunk 13 above)
Comment (yosefe) [reply src/uct/ib/rdmacm/rdmacm_listener.c:69]: also, what about tcp sockcm?

Comment (evgeny-leksikov) [reply src/uct/ib/rdmacm/rdmacm_listener.c:69]: > also, what about tcp sockcm?

I dont think that tcp has such issue, if the interface has an IP address TCP should be abale to work with it, that is not true for rdmacm

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@evgeny-leksikov can u pls squash?

**Comment by evgeny-leksikov** [comment]
> @evgeny-leksikov can u pls squash?

done

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@shamisp @tonycurtis can you pls review the API doc change?

**Review by tonycurtis — COMMENTED** [review]
ucp.h:2189: "should" or "must"?

**Review by yosefe — APPROVED** [review]

File: src/ucp/core/ucp_listener.c [hunk 15]
```diff
@@ -152,69 +166,81 @@ ucp_listen(ucp_listener_h listener, const ucp_listener_params_t *params)
...
-        }
-
-        uct_attr.field_mask = UCT_LISTENER_ATTR_FIELD_SOCKADDR;
-        status = uct_listener_query(uct_listeners[listener->num_rscs - 1],
-                                    &uct_attr);
-        if (status != UCS_OK) {
-            goto err_destroy_listeners;
-        }
+            uct_attr.field_mask = UCT_LISTENER_ATTR_FIELD_SOCKADDR;
+            status = uct_listener_query(uct_listeners[listener->num_rscs - 1],
```
Comment (dmitrygx) [inline src/ucp/core/ucp_listener.c:183]: minor: should we align by `=`?

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 16]
```diff
@@ -815,6 +826,72 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle_without_err_cb)
...
+    /* define both conn and accept handlers to valid callbacks
+     * (should be only 1) */
+    params.field_mask        = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR |
+                               UCP_LISTENER_PARAM_FIELD_CONN_HANDLER |
+                               UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER;
+    params.conn_handler.cb   =
+            (ucp_listener_conn_callback_t)ucs_empty_function;
+    params.accept_handler.cb =
+            (ucp_listener_accept_callback_t)ucs_empty_function;
+    status = create_listener_wrap_err(params, listener);
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_sockaddr.cc:877]: minor: align by `=`?

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 17]
```diff
@@ -815,6 +826,72 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle_without_err_cb)
...
+    EXPECT_EQ(UCS_ERR_INVALID_PARAM, status);
+
+    /* define both conn and accept handlers to NULL */
+    params.field_mask         = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR |
+                                UCP_LISTENER_PARAM_FIELD_CONN_HANDLER |
+                                UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER;
+    params.accept_handler.cb  = NULL;
+    params.accept_handler.arg = NULL;
+    status = create_listener_wrap_err(params, listener);
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_sockaddr.cc:865]: minor: align by `=`?

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 18]
```diff
@@ -815,6 +826,72 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle_without_err_cb)
...
+    /* accept handlers aren't set */
+    status = create_listener_wrap_err(params, listener);
+    EXPECT_EQ(UCS_ERR_INVALID_PARAM, status);
+
+    /* define conn handler flag but set to NULL */
+    params.field_mask       = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR |
+                              UCP_LISTENER_PARAM_FIELD_CONN_HANDLER;
+    params.conn_handler.cb  = NULL;
+    params.conn_handler.arg = NULL;
+    status = create_listener_wrap_err(params, listener);
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_sockaddr.cc:856]: minor: align by `=`?

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 19]
```diff
@@ -815,6 +826,72 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle_without_err_cb)
...
+            (ucp_listener_conn_callback_t)ucs_empty_function;
+    params.accept_handler.cb =
+            (ucp_listener_accept_callback_t)ucs_empty_function;
+    status = create_listener_wrap_err(params, listener);
+    EXPECT_EQ(UCS_ERR_INVALID_PARAM, status);
+
+    /* sockaddr and valid conn handler is OK */
+    params.field_mask = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR |
+                        UCP_LISTENER_PARAM_FIELD_CONN_HANDLER;
+    status = create_listener_wrap_err(params, listener);
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_sockaddr.cc:883]: minor: align by `=`?

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 20]
```diff
@@ -815,6 +826,72 @@ UCS_TEST_P(test_ucp_sockaddr, err_handle_without_err_cb)
...
+    ASSERT_UCS_OK(status);
+    ucp_listener_destroy(listener);
+
+    /* sockaddr and valid accept handler is OK */
+    params.field_mask = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR |
+                        UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER;
+    status = create_listener_wrap_err(params, listener);
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_sockaddr.cc:890]: minor: align by `=`?

**Review by dmitrygx — APPROVED** [review]

**Review by dmitrygx — APPROVED** [review]

**Review by yosefe — APPROVED** [review]
